### PR TITLE
config(slicing): add logical partitioning map

### PR DESCRIPTION
**Pull request by onewland**
This adds logical partitioning configuration to Sentry, mirroring what exists in Snuba.

References:
- https://github.com/getsentry/snuba/blob/master/snuba/datasets/partitioning.py
- https://getsentry.github.io/snuba/architecture/partitioning.html
- (internal) https://www.notion.so/sentry/Generic-Metrics-Slicing-9c206e02f92a4abca5948d2e21c6c00d

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -2864,3 +2864,11 @@ def build_cdc_postgres_init_db_volume(settings):
...
 # How long we should wait for a gateway proxy request to return before giving up
 GATEWAY_PROXY_TIMEOUT = None
+
+SENTRY_SLICING_LOGICAL_PARTITION_COUNT = 256
+# This maps a dataset for slicing by name and (lower logical partition, upper physical partition)
+# to a given slice. A slice is a set of physical resources in Sentry and Snuba.
+#
+# For each dataset, the range [0, SENTRY_SLICING_LOGICAL_PARTITION_COUNT) must be mapped
+# to a slice ID
+SENTRY_SLICING_CONFIG: Mapping[str, Mapping[Tuple[int, int], int]] = {}
```
Comment (nikhars) [inline src/sentry/conf/server.py:2874]: I don't know if dataset is the right terminology for Sentry. It feels more like a storage term. What if we state these mappings in terms of entities or something else?

File: src/sentry/ingest/partitioning.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    for ((logical_part_lo_incl, logical_part_hi_excl), slice_id) in settings.SENTRY_SLICING_CONFIG[
+        dataset_name
+    ].items():
+        if logical_partition >= logical_part_lo_incl and logical_partition < logical_part_hi_excl:
+            return slice_id
+
+    assert False, f"no mapping for dataset {dataset_name} for logical_partition {logical_partition}"
+
+
+def is_storage_partitioned(dataset_name: str) -> bool:
```
Comment (nikhars) [inline src/sentry/ingest/partitioning.py:38]: Based on my previous comment, now I definitely feel storage in not the right name. You could be partitioning any resource in Sentry like consumers, redis etc.

Comment (onewland) [reply src/sentry/ingest/partitioning.py:38]: Yeah, I think unfortunately there isn't a good cross-system term for this. What do you think of

```python
type Sliceable = Literal["generic_metrics_sets", "generic_metrics_distributions"]

def map_org_id_to_logical_partition(org_id: int) -> int: # no change
  ...

def map_logical_partition_to_slice(sliceable: Sliceable, logical_partition: int) -> int:
  ...

def is_storage_partitioned(sliceable: Sliceable) -> bool:
  ...
```

Comment (onewland) [reply src/sentry/ingest/partitioning.py:38]: maybe the last one could just be named `is_sliced`

Comment (nikhars) [reply src/sentry/ingest/partitioning.py:38]: > Yeah, I think unfortunately there isn't a good cross-system term for this. What do you think of
> 
> ```python
> Sliceable = Literal["generic_metrics_sets", "generic_metrics_distributions"]
> 
> def map_org_id_to_logical_partition(org_id: int) -> int: # no change
>   ...
> 
> def map_logical_partition_to_slice(sliceable: Sliceable, logical_partition: int) -> int:
>   ...
> 
> def is_sliceable_partitioned(sliceable: Sliceable) -> bool:
>   ...
> ```
Yes that is better

**Comment by onewland** [comment]
> > for when ingest team wakes up: my understanding is that those settings are only supposed to be affecting the metrics indexer. @onewland: `sentry.ingest` is currently more closely related to relay and pushing config into relay (which i understand has nothing to do with this PR)
> > since this only affects metrics indexer i would move this code into `sentry.sentry_metrics`. At the same time I understand that this code is probably not metrics-specific. I guess a new top-level module, or extending the scope of `sentry.ingest` is both fime
> 
> Also, I don't know how much it would matter if we need to import these into sentry metrics indexer. There are certain requirements which need to be made when having configuration for the indexer. Please see https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/configuration.py#L1-L5

Do you mean this stuff *should* move into sentry_metrics? I don't feel like it belongs there long-term but I don't want to go in circles with this discussion -- it can move later

**Review by nikhars — APPROVED** [review]

**Comment by nikhars** [comment]
> > > for when ingest team wakes up: my understanding is that those settings are only supposed to be affecting the metrics indexer. @onewland: `sentry.ingest` is currently more closely related to relay and pushing config into relay (which i understand has nothing to do with this PR)
> > > since this only affects metrics indexer i would move this code into `sentry.sentry_metrics`. At the same time I understand that this code is probably not metrics-specific. I guess a new top-level module, or extending the scope of `sentry.ingest` is both fime
> > 
> > 
> > Also, I don't know how much it would matter if we need to import these into sentry metrics indexer. There are certain requirements which need to be made when having configuration for the indexer. Please see https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/configuration.py#L1-L5
> 
> Do you mean this stuff _should_ move into sentry_metrics? I don't feel like it belongs there long-term but I don't want to go in circles with this discussion -- it can move to a better home later

It might not be the ideal place. But it might have to be put there because of some restrictions. Its not a blocker for this PR but might have to be addressed later on

File: src/sentry/ingest/partitioning.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    return org_id % settings.SENTRY_SLICING_LOGICAL_PARTITION_COUNT
+
+
+def map_logical_partition_to_slice(sliceable: Sliceable, logical_partition: int) -> int:
+    """
+    Maps a logical partition to a slice.
+    """
+
+    assert is_sliced(sliceable), f"cannot retrieve slice of non-partitioned sliceable {sliceable}"
```
Comment (jjbayer) [inline src/sentry/ingest/partitioning.py:26]: nit: This assert seems redudant to me. It resolves to `assert sliceable in settings.SENTRY_SLICING_CONFIG`, which is also what the next line asserts.

**Review by jjbayer — APPROVED** [review]
Agree with @untitaker that this maybe should not live in `sentry/ingest` (see codeowners), but that should not block the merge.

File: src/sentry/ingest/partitioning.py (hunk 3 above)
Comment (onewland) [reply src/sentry/ingest/partitioning.py:26]: you're right, good catch

**Pull request merged**